src/devices/src/virtio/queue.rs

```diff
@@ -300,7 +300,21 @@ impl Queue {
 
     /// Pop the first available descriptor chain from the avail ring.
     pub fn pop<'a, 'b>(&'a mut self, mem: &'b GuestMemoryMmap) -> Option<DescriptorChain<'b>> {
-        if self.len(mem) == 0 {
+        let len = self.len(mem);
+        // The number of descriptor chain heads to process should always
+        // be smaller or equal to the queue size, as the driver should
+        // never ask the VMM to process a available ring entry more than
+        // once. Checking and reporting such incorrect driver behavior
+        // can prevent potential hanging and Denial-of-Service from
+        // happening on the VMM side.
+        if len > self.actual_size() {
+            // We are choosing to interrupt execution since this could be a potential malicious driver
+            // scenario. This way we also eliminate the risk of repeatedly logging and potentially
+            // clogging the microVM through the log system.
+            panic!("The number of available virtio descriptors is greater than queue size!");
+        }
```

Comment:
```suggestion
        // We are choosing to interrupt execution since this could be a potential malicious driver
        // scenario. This way we also eliminate the risk of repeatedly logging and potentially
        // clogging the microVM through the log system.
        assert!(len > self.actual_size(), "The number of available virtio descriptors is greater than queue size!");
```

Reply:
> At least, this is what we have targeted to do until now so that it is easier to track the panic branches we have in production code.

Could you expand, how do we currently track panic branches?

> Also, we have a panic handler installed when starting Firecracker (for flushing metrics and whatnot) which I am not sure would get called if we use asssert.

The docs for [`std::assert`](https://doc.rust-lang.org/std/macro.assert.html) specify:

> This will invoke the [panic!](https://doc.rust-lang.org/core/macro.panic.html) macro if the provided expression cannot be evaluated to true at runtime.

Reply:
>  At least, this is what we have targeted to do until now so that it is easier to track the panic branches we have in production code.
> 
>> Could you expand, how do we currently track panic branches?
>
We do not have an automated process or something of the kind. But if we were to use `assert`s in production code (which are extensively used throughout unit tests) is going to be quite challenging to uncover those if the need arises. That would be one of the practical aspects of it.
> > Also, we have a panic handler installed when starting Firecracker (for flushing metrics and whatnot) which I am not sure would get called if we use asssert.
> 
> The docs for [`std::assert`](https://doc.rust-lang.org/std/macro.assert.html) specify:
> 
> > This will invoke the [panic!](https://doc.rust-lang.org/core/macro.panic.html) macro if the provided expression cannot be evaluated to true at runtime.

Nice, thanks for looking this up!
Can you tell me is there is a strong reason for using assert over panic here?






> I prefer to use panic over assert in production code. At least, this is what we have targeted to do until now so that it is easier to track the panic branches we have in production code. Also, we have a panic handler installed when starting Firecracker (for flushing metrics and whatnot) which I am not sure would get called if we use asssert.

Reply:
It is simpler, but beyond that, no strong reasons.